sdk/rust/oak_derive/src/lib.rs

```diff
@@ -59,24 +58,22 @@ pub fn derive_oak_exports(input: TokenStream) -> TokenStream {
 
     let expanded = quote! {
         #[no_mangle]
-        pub extern "C" fn oak_main(handle: u64) -> i32 {
+        pub extern "C" fn oak_main(handle: u64) {
             // A panic in the Rust module code cannot safely pass through the FFI
-            // boundary, so catch any panics here and translate to an error return.
+            // boundary, so catch any panics here and drop them.
             // https://doc.rust-lang.org/nomicon/ffi.html#ffi-and-panics
-            std::panic::catch_unwind(||{
+            let _ = std::panic::catch_unwind(||{
                 oak::set_panic_hook();
                 inner_main(handle)
-            })
-            .unwrap_or(Err(oak::OakStatus::ERR_INTERNAL))
-            .err()
-            .unwrap_or(oak::OakStatus::OK)
-            .value()
+            });
         }
         // Internal version of the main entrypoint, to allow testing without any
         // panic interception.
-        pub fn inner_main(handle: u64) -> Result<(), oak::OakStatus> {
+        pub fn inner_main(handle: u64) {
             let mut node = <#name>::new();
-            oak::grpc::event_loop(node, oak::ReadHandle{ handle: oak::Handle::from_raw(handle) })
+            if let Err(s) = oak::grpc::event_loop(node, oak::ReadHandle{ handle: oak::Handle::from_raw(handle) }) {
+                warn!("Node terminating with {:?}", s);
+            }
```

Comment:
I'm still not convinced this is easier to read or maintain than leaving it as `pub fn inner_main(handle: u64) -> Result<(), oak::OakStatus>` (or in fact change it to `Result<(), Box<dyn Debug>>`, since all we do with the error is log it, and this seems more semantically accurate), and just unwrapping the error on line 67, and letting the panic hook deal with it.

Reply:
My aim here was to have the `inner_main` is as close to the "real" entrypoint as possible (the only differences being no FFI-required `catch_unwind`, no `extern "C"`).  As [previously mentioned](https://github.com/project-oak/oak/pull/475#discussion_r367936178), this removes the need for the (test) runtime to deal with Rust types (rather than pure Wasm types), which would be needed for a fully Rust-based runtime later.

To put it another way: `NodeMain` moving into `oak_abi` means that it has to have a Wasm-compatible signature, and splitting out `inner_main` was specifically done to expose a `NodeMain` entrypoint for testing.

Reply:
But this still leaves a gap between the two, in that the inner function may panic, while the outer function may not (because of the `catch_unwind`). So you still need to deal with this difference. Anyways, it's fine as it is for now, but I think it would be nice to have clear requirements for the inner and outer functions at some point (e.g. no panics crossing outer function boundary, etc.), and come up with an ergonomic API that satisfies it.